File: android/ValleTPV/app/src/main/java/com/valleapp/valletpv/Camareros.java
/*
  @Author: Manuel Rodriguez <valle>
 * @Date:   2018-12-29T11:26:27+01:00
 * @Email:  valle.mrv@gmail.com
 * @Last modified by:   valle
 * @Last modified time: 2018-12-29T14:52:12+01:00
 * @License: Apache License v2.0
 */
package com.valleapp.valletpv;

import android.app.Activity;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.Bundle;
import android.os.Handler;
import android.os.IBinder;
import android.os.Looper;
import android.os.Message;
import android.util.DisplayMetrics;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Button;
import android.widget.TableLayout;
import android.widget.TableRow;
import android.widget.Toast;

import com.valleapp.valletpv.db.DBCamareros;
import com.valleapp.valletpv.tools.ServicioCom;
import com.valleapp.valletpv.tools.ToastShowInfoCuenta;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;


public class Camareros extends Activity {


    final Context cx = this;

    JSONArray lscam = null;
    JSONObject cam_sel = null;
    ServicioCom myServicio;

    DBCamareros dbCamareros;

    int presBack = 0;

   ToastShowInfoCuenta toastShowInfoCuenta;

    private final Handler handleHttp = new Handler(Looper.getMainLooper()) {
        public void handleMessage(Message msg) {
            String op = msg.getData().getString("op");
            if (op == null) {
                rellenarCamareros();
            }else{
                if (op == "show_info_cobro"){
                    Bundle datos = msg.getData();
                    Double entrega = datos.getDouble("entrega");
                    Double cambio = datos.getDouble("cambio");
                    if (cambio > 0) {
                        LayoutInflater inflater = getLayoutInflater();
                        View layout = inflater.inflate(R.layout.toast_info_cambio, findViewById(R.id.toast_info_cobro_container));
                        if (toastShowInfoCuenta != null) toastShowInfoCuenta.cancel();
                        toastShowInfoCuenta = new ToastShowInfoCuenta();
                        toastShowInfoCuenta.show(entrega, cambio, 10000, getApplicationContext(), layout);
                    }
                }
            }
        }
    };

    private void rellenarCamareros() {
       try {
           lscam = dbCamareros.filter("autorizado=1");
           if(lscam.length()>0){

               TableLayout ll = findViewById(R.id.pneCamareros);
               ll.removeAllViews();

               DisplayMetrics metrics = getResources().getDisplayMetrics();


               TableLayout.LayoutParams params = new TableLayout.LayoutParams(
                          TableLayout.LayoutParams.MATCH_PARENT,
                          TableLayout.LayoutParams.WRAP_CONTENT);


               TableRow.LayoutParams rowparams = new TableRow.LayoutParams(
                          TableLayout.LayoutParams.MATCH_PARENT,
                          Math.round(metrics.density * 120));

               rowparams.setMargins(5,5,5,5);
               ll.setStretchAllColumns(true);

               TableRow row = new TableRow(cx);

               ll.addView(row, params);


               for (int i = 0; i < lscam.length(); i++) {
                   JSONObject  cam =  lscam.getJSONObject(i);
                   Button btn = new Button(cx);
                   btn.setId(i);
                   btn.setSingleLine(false);
                   String[] nom = cam.getString("Nombre").split(" ");

                   btn.setText(nom.length > 1 ? nom[0]+"\n"+nom[1] : cam.getString("Nombre") +"\n");
                   btn.setBackgroundResource(R.drawable.fondo_btn_xml);
                   btn.setOnClickListener(view -> {
                       try {
                           final JSONObject obj =  lscam.getJSONObject(view.getId());
                           entrar_en_mesas((obj));
                       } catch (JSONException e) {
                           e.printStackTrace();
                       }
                      });
                      row.addView(btn, rowparams);

                      if (((i+1) % 6) == 0) {
                          row = new TableRow(cx);
                          ll.addView(row, params);
                      }
               }
           }

       }catch (Exception e){
          e.printStackTrace();
       }
    }

    protected void entrar_en_mesas(final JSONObject camarero) throws JSONException {
        presBack = 0;
        cam_sel = camarero;
        Intent intent = new Intent(cx, Mesas.class);
        intent.putExtra("cam", cam_sel.toString());
        startActivity(intent);
    }


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_camareros);
    }


    @Override
    public void onBackPressed() {
        if(presBack>=1) {
            super.onBackPressed();
        }else{
            Toast.makeText(getApplicationContext(),"Pulsa otra vez para salir", Toast.LENGTH_SHORT).show();
            presBack++;
        }
    }

    @Override
    protected void onResume() {
        presBack = 0;
        Intent intent = new Intent(getApplicationContext(), ServicioCom.class);
        bindService(intent, mConexion, Context.BIND_AUTO_CREATE);
        if (myServicio != null){
            rellenarCamareros();
        }
        super.onResume();
    }

    @Override
    protected void onDestroy() {
        unbindService(mConexion);
        super.onDestroy();
    }

    private ServiceConnection mConexion = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName componentName, IBinder iBinder) {
            myServicio = ((ServicioCom.MyBinder)iBinder).getService();
            if(myServicio!=null){
                myServicio.setExHandler("camareros", handleHttp);
                dbCamareros = (DBCamareros) myServicio.getDb("camareros");
                rellenarCamareros();
            }

        }

        @Override
        public void onServiceDisconnected(ComponentName componentName) {
            myServicio = null;
        }
    };


}

File: android/ValleTPV/ValleTPV/src/main/java/com/valleapp/valletpv/Camareros.kt
package com.valleapp.valletpv

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.Bundle
import android.os.Handler
import android.os.IBinder
import android.os.Looper
import android.util.Log
import android.widget.Button
import android.widget.TableLayout
import android.widget.TableRow
import android.widget.Toast
import androidx.activity.OnBackPressedCallback
import androidx.appcompat.app.AppCompatActivity
import com.tu.paquete.CustomToast
import com.valleapp.valletpv.tools.ServiceTPV
import com.valleapp.valletpv.tools.ToastShowInfoCuenta
import com.valleapp.valletpvlib.db.DBCamareros
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import kotlin.math.roundToInt

class Camareros : AppCompatActivity() {

    private val cx: Context = this
    private var lscam: JSONArray? = null
    private var camSel: JSONObject? = null
    private var myServicio: ServiceTPV? = null
    private var dbCamareros: DBCamareros? = null
    private var presBack = 0
    private var toastShowInfoCuenta: ToastShowInfoCuenta? = null
    private val customToast = CustomToast(this)
    private var server = ""

    private lateinit var onBackPressedCallback: OnBackPressedCallback

    private val handleHttp = Handler(Looper.getMainLooper()) { msg ->
        val op = msg.data.getString("op")
        
        if (op == null) {
            rellenarCamareros()
        } else {
            if (op == "show_info_cobro") {
                val datos = msg.data
                val entrega = datos.getDouble("entrega")
                val cambio = datos.getDouble("cambio")
                val inflater = layoutInflater
                val layout = inflater.inflate(R.layout.toast_info_cambio, findViewById(R.id.toast_info_cobro_container))
                toastShowInfoCuenta?.cancel()
                toastShowInfoCuenta = ToastShowInfoCuenta()
                if (entrega > 0) {
                    toastShowInfoCuenta?.show(entrega, cambio, 10000, applicationContext, layout)
                } else {
                    toastShowInfoCuenta?.showMessageOnly("Cobro aceptado\ncon tarjeta", 10000, applicationContext, layout)
                }
            }
        }
        true
    }

    private fun rellenarCamareros() {
        try {
            lscam = dbCamareros?.filter("autorizado=1")
            if ((lscam?.length() ?: 0) > 0) {
                val ll = findViewById<TableLayout>(R.id.pneCamareros)
                ll.removeAllViews()

                val metrics = resources.displayMetrics

                val params = TableLayout.LayoutParams(
                    TableLayout.LayoutParams.MATCH_PARENT,
                    TableLayout.LayoutParams.WRAP_CONTENT
                )

                val rowparams = TableRow.LayoutParams(
                    TableRow.LayoutParams.MATCH_PARENT,
                    (metrics.density * 120).roundToInt()
                ).apply {
                    setMargins(5, 5, 5, 5)
                }

                ll.isStretchAllColumns = true

                var row = TableRow(cx)
                ll.addView(row, params)

                for (i in 0 until lscam!!.length()) {
                    val cam = lscam!!.getJSONObject(i)
                    val btn = Button(cx).apply {
                        id = i
                        isSingleLine = false
                        text = String.format("%s\n%s", cam.getString("nombre"), cam.getString("apellidos"))
                        setBackgroundResource(com.valleapp.valletpvlib.R.drawable.fondo_btn_xml)
                        textSize = 20f // Establecer el tamaño del texto a 26sp
                        setOnClickListener {
                            try {
                                val obj = lscam!!.getJSONObject(id)
                                entrarEnMesas(obj)
                            } catch (_: JSONException) {
                                Log.e("ERROR", "Error al acceder al camarero")
                            }
                        }
                    }
                    row.addView(btn, rowparams)

                    if ((i + 1) % 5 == 0) {
                        row = TableRow(cx)
                        ll.addView(row, params)
                    }
                }
            }
        } catch (_: Exception) {
            Log.e("ERROR", "Error al acceder al camarero")
        }
    }

    private fun entrarEnMesas(camarero: JSONObject) {
        presBack = 0
        camSel = camarero
        val intent = Intent(cx, Mesas::class.java).apply {
            putExtra("cam", camSel.toString())
            putExtra("server", server)
        }
        startActivity(intent)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_camareros)
        server = intent.getStringExtra("server") ?: ""
        onBackPressedCallback = object : OnBackPressedCallback(true) {
            override fun handleOnBackPressed() {
                if (presBack >= 1) {
                    finish() // Cierra la actividad
                } else {
                    customToast.showBottom(
                        "Pulsa otra vez para salir",
                        Toast.LENGTH_SHORT
                    )
                    presBack++
                }
            }
        }
        onBackPressedDispatcher.addCallback(this, onBackPressedCallback)
    }



    override fun onResume() {
        super.onResume()
        presBack = 0
        val intent = Intent(applicationContext, ServiceTPV::class.java)
        bindService(intent, mConexion, BIND_AUTO_CREATE)
        rellenarCamareros()
    }

    override fun onDestroy() {
        unbindService(mConexion)
        onBackPressedCallback.remove()
        super.onDestroy()
    }

    private val mConexion = object : ServiceConnection {
        override fun onServiceConnected(componentName: ComponentName, iBinder: IBinder) {
            myServicio = (iBinder as ServiceTPV.MyBinder).service
            myServicio?.setExHandler("camareros", handleHttp)
            dbCamareros = myServicio?.getDb("camareros") as? DBCamareros
            rellenarCamareros()
        }

        override fun onServiceDisconnected(componentName: ComponentName) {
            myServicio = null
        }
    }
}